app/assets/javascripts/provider/admin/apiconfig/services/service_discovery.js

```diff
@@ -0,0 +1,122 @@
+  const formDiscover = document.getElementById('service_discovery');
+
+  function clearSelectOptions(select) {
+    const options = select.getElementsByTagName('option');
+    for (var i = options.length -1; i > 0; i--) {
+      options[i].remove();
+    }
+  };
+
+  function clearNamespaces() {
+    return clearSelectOptions(formDiscover.service_namespace);
+  };
+
+  function clearServiceNames() {
+    return clearSelectOptions(formDiscover.service_name);
+  };
+
+  function showAndHideForms(showForm, hideForm) {
+    showForm.classList.remove('is-hidden');
+    hideForm.classList.add('is-hidden');
+  }
+
+  function showServiceForm(source, discoverFunc, scratchFunc) {
+    if (source === 'discover') {
+      showAndHideForms(formDiscover, formScratch);
+      return typeof discoverFunc === "function" ? discoverFunc() : undefined;
+    } else {
+      showAndHideForms(formScratch, formDiscover);
+      return typeof scratchFunc === "function" ? scratchFunc() : undefined;
+    }
+  };
+
+  function changeServiceSource() {
+    clearNamespaces();
+    clearServiceNames();
+    return showServiceForm(formSource.source.value, fetchNamespaces);
+  };
+
+  function changeClusterNamespace() {
+    clearServiceNames();
+    const selectedNamespace = formDiscover.service_namespace.value;
+    return selectedNamespace !== '' ? fetchServices(selectedNamespace) : undefined;
+  };
+
+  function addOptionToSelect(selectElem, val) {
+    var opt = document.createElement('option');
+    opt.text = val;
+    return selectElem.appendChild(opt);
+  };
+
+  function fetchData(url, type) {
+    if ('fetch' in window) {
+      fetch(url)
+      .then(function(response) {
+          return response.json();
+      })
+      .then(function(data){
+        populateOptions(type, data);
+      });
+    } else {
+      return $.getJSON(url, function(data) {
+        populateOptions(type, data);
+      });
+    }
+  }
+
+  function populateOptions(type, data) {
+    switch (type) {
+      case 'namespaces':
+        populateNamespaces(data);
+        break;
+      case 'services':
+        populateServices(data);
+        break;
+    }
+  }
+
+  function populateNamespaces(data) {
+    const projects = data.projects;
+    projects.forEach(function(project, index, array){
+      addOptionToSelect(formDiscover.service_namespace, project.metadata.name);
+    });
+    return projects.length > 0 ? (
+      formDiscover.service_namespace.selectedIndex = 1,
+      changeClusterNamespace()
+    ) : undefined;
+  }
+
+  function fetchNamespaces() {
+    const URLNamespaces = '/p/admin/service_discovery/projects.json';
+    fetchData(URLNamespaces, 'namespaces');
+  }
+
+  function populateServices(data) {
+    const services = data.services;
+    services.forEach(function(service, index, array) {
+      addOptionToSelect(formDiscover.service_name, service.metadata.name);
+    });
+    return services.length > 0 ? formDiscover.service_name.selectedIndex = 1 : undefined;
```

Comment:
This line is confusing, do we need a return value?

Reply:
And also https://github.com/3scale/porta/pull/110/files#diff-5b5230bd99ca1dc0c8d7656f9992206bR87